src/middle/Mir.ml

```diff
@@ -164,26 +157,39 @@ type ('e, 's) prog =
   ; output_vars: 'e io_var list
   ; prog_name: string
   ; prog_path: string }
-[@@deriving sexp]
+[@@deriving sexp, map]
 
-type 'm with_expr = {expr: 'm with_expr expr; emeta: 'm} [@@deriving sexp]
+type 'm with_expr = {expr: 'm with_expr expr; emeta: 'm}
+[@@deriving compare, sexp, hash]
 
 type mtype_loc_ad =
   { mtype: unsizedtype
   ; mloc: location_span sexp_opaque [@compare.ignore]
   ; madlevel: autodifftype }
-[@@deriving sexp]
+[@@deriving compare, sexp, hash]
 
 type ('e, 'm) stmt_with =
   {stmt: ('e with_expr, ('e, 'm) stmt_with) statement; smeta: 'm}
 [@@deriving sexp]
 
+type ('e, 'm) stmt_with_num = {stmtn: ('e with_expr, int) statement; smetan: 'm}
+[@@deriving sexp, hash]
+
+type expr_no_meta = unit with_expr
+
+type expr_typed_located = mtype_loc_ad with_expr
+[@@deriving sexp, compare, hash]
+
+type stmt_no_meta = (expr_no_meta, unit) stmt_with
+
 type stmt_loc =
   (mtype_loc_ad, (location_span sexp_opaque[@compare.ignore])) stmt_with
 [@@deriving sexp]
 
-type expr_no_meta = unit with_expr [@@deriving sexp]
-type stmt_no_meta = (expr_no_meta, unit) stmt_with [@@deriving sexp]
+type stmt_loc_num =
```

Comment:
Should the num variants be left inside the optimizer? Not sure if y'all envision use cases by third parties.

Reply:
I'll move them to the optimizer.